### PR TITLE
Various partitioning method for parallel::shared::Triangulation

### PR DESCRIPTION
**Pull request by tcclevenger**
- Implementation of a zorder partition (what is used in p4est), multigrid level partition ("youngest child" rule used in distributed::Triangulation), and the option for the user to define a custom partition
- Adds tests for each partitioning scheme listed above.
- Implements level versions of two functions in grid_tools (get_vertex_connectivity_of_cells_on_level, compute_cell_halo_layer_on_level) that were needed for the implementation of the multigrid partitioning

### CODE REVIEW

File: include/deal.II/distributed/shared_tria.h [hunk 1]
```diff
@@ -72,6 +72,58 @@ namespace parallel
       typedef typename dealii::Triangulation<dim,spacedim>::cell_iterator        cell_iterator;
 
       /**
+       * Configuration flags for distributed Triangulations to be set in the
+       * constructor. Settings can be combined using bitwise OR.
```
Comment (davydden) [inline include/deal.II/distributed/shared_tria.h]: maybe say explicitly that `construct_multigrid_hierarchy` can be combined with other options. Otherwise your assert below says they can not be combined arbitrarily.

File: source/distributed/shared_tria.cc [hunk 2]
```diff
@@ -36,17 +36,39 @@ namespace parallel
...
 
 
     template <int dim, int spacedim>
     void Triangulation<dim,spacedim>::partition()
     {
-      dealii::GridTools::partition_triangulation (this->n_subdomains, *this);
+      if (settings & partition_metis)
+        {
+          dealii::GridTools::partition_triangulation (this->n_subdomains, *this);
+          if (settings & construct_multigrid_hierarchy)
```
Comment (davydden) [inline source/distributed/shared_tria.cc]: if this applies to all, maybe move it outside of the first `if`, namely
```
if (settings & partition_metis)
{
}
else if(settings & partition_zorder)
{
}
else if(settings & partition_custom_signal)
{
}
AssertThrow(false, ExcInternalError())

if (settings & construct_multigrid_hierarchy)
  //do something additionally
```

File: source/grid/grid_tools.cc [hunk 3]
```diff
@@ -1606,6 +1606,45 @@ namespace GridTools
   }
 
 
+  template <class MeshType>
+  std::vector<typename MeshType::cell_iterator>
+  compute_cell_halo_layer_on_level
```
Comment (davydden) [inline source/grid/grid_tools.cc]: i think it would be better if you
1. try to re-use  `compute_active_cell_halo_layer()` (maybe extend it or maybe move the implementation inside internal function and call it from two places). Either way, it look slike there is a fair amount of code duplication here. There only difference is `begin_active()/end()` vs `begin(level) /end(level)`.
2. add a predicate here `const std_cxx11::function<bool (const typename MeshType::active_cell_iterator &)> &predicate` instead of hard-coding comparison of subdomain id.

File: source/grid/grid_tools.cc [hunk 4]
```diff
@@ -1629,6 +1668,7 @@ namespace GridTools
 
 
 
+
```
Comment (davydden) [inline source/grid/grid_tools.cc]: we usually keep 3 lines between definitions, no need for an extra line here.

**Comment by tjhei** [comment]
I guess it is worth mentioning that:
1. This is the first PR to get towards GMG on ``shared::Tria`` but this is not part of this PR. We can now partition the cells for GMG now.
2. This partly addresses #1364. If we wanted, we could now run all tests that are done for ``p::d`` also with ``shared::Tria``.

File: source/distributed/shared_tria.cc (hunk 2 above)
Comment (tjhei) [inline source/distributed/shared_tria.cc]: I think you are missing an ``else`` before ``AssertThrow``, Denis.

Comment (tjhei) [inline source/distributed/shared_tria.cc]: Conrad, keep in mind that we don't want to do the multigrid assignment for ``partition_custom_signal``.

**Comment by tamiko** [comment]
I have added the "WIP" label (even though it looks like only minor modifications are required) and added the pull request to the 9.0 milestone :-)

File: include/deal.II/distributed/shared_tria.h [hunk 5]
```diff
@@ -72,6 +72,58 @@ namespace parallel
...
+         *     void mypartition(parallel::shared::Triangulation<dim> &tria)
+         *     {
+         *       // user defined partitioning scheme
+         *     }
+         *
+         *     int main ()
+         *     {
+         *       parallel::shared::Triangulation<dim> tria(...,
+         *                                                 parallel::shared::Triangulation<dim>::Settings::partition_custom_signal);
+         *       shared_tria.signals.post_refinement.connect (std::bind(&mypartition<dim>, std::ref(shared_tria)));
```
Comment (bangerth) [inline include/deal.II/distributed/shared_tria.h]: The object is actually only called `tria` in your example, not `shared_tria`.

File: include/deal.II/distributed/shared_tria.h [hunk 6]
```diff
@@ -72,6 +72,58 @@ namespace parallel
...
+         *     }
+         *  @endcode
+         */
+        partition_custom_signal = 0x4,
+
+        /**
+         * This flags needs to be set to use the geometric multigrid
```
Comment (bangerth) [inline include/deal.II/distributed/shared_tria.h]: this flags -> this flag

File: include/deal.II/distributed/shared_tria.h [hunk 7]
```diff
@@ -72,6 +72,58 @@ namespace parallel
...
+         * functionality. This option requires additional computation and
+         * communication. Note: geometric multigrid is still a work in
+         * progress and not yet functional for shared triangulation.
```
Comment (bangerth) [inline include/deal.II/distributed/shared_tria.h]: this is the sort of comment we will forget to remove when the MG actually starts working. I would just omit the note.

**Comment by bangerth** [comment]
On 03/15/2017 01:40 PM, Timo Heister wrote:
> # This is the first PR to get towards GMG on |shared::Tria| but this is not part
> of this PR. We can now partition the cells for GMG now.

Yay. I think that would be useful to give p::s::Tria better test coverage. Not 
necessary as part of the current pull request, but it would be nice if that 
was one of the next steps.

File: source/distributed/shared_tria.cc [hunk 8]
```diff
@@ -82,7 +104,78 @@ namespace parallel
...
+        AssertThrowMPI(ierr);
+
+        Assert(total_cells == this->n_active_cells(),
+               ExcMessage("Not all cells are assigned to a processor."))
+      }
+#endif
 
+      if (settings & construct_multigrid_hierarchy)
+        {
+#ifdef DEBUG
```
Comment (tjhei) [inline source/distributed/shared_tria.cc]: can you move this ``ifdef`` outside the ``if``?

File: source/grid/grid_tools.cc [hunk 9]
```diff
@@ -2125,6 +2191,155 @@ namespace GridTools
...
+             cell = triangulation.begin_active();
+             cell != triangulation.end(); ++cell)
+          cell->set_subdomain_id (0);
+        return;
+      }
+
+    typedef typename Triangulation<dim,spacedim>::active_cell_iterator ac_it;
+    typedef typename Triangulation<dim,spacedim>::cell_iterator lvl_it;
+
+    unsigned int current_proc_idx=0;
```
Comment (tjhei) [inline source/grid/grid_tools.cc]: move this down to where you first need it.

File: source/grid/grid_tools.cc [hunk 10]
```diff
@@ -2125,6 +2191,155 @@ namespace GridTools
...
+    unsigned int current_cell_idx=0;
+    const unsigned int n_active_cellls = triangulation.n_active_cells();
```
Comment (tjhei) [inline source/grid/grid_tools.cc]: type in ``cellls``

File: source/grid/grid_tools.cc [hunk 11]
```diff
@@ -2125,6 +2191,155 @@ namespace GridTools
...
+
+    // create coarse cell reordering
```
Comment (tjhei) [inline source/grid/grid_tools.cc]: add a comment that that we need to duplicate the ordering from p4est.

File: source/grid/grid_tools.cc [hunk 12]
```diff
@@ -2125,6 +2191,155 @@ namespace GridTools
...
+        const unsigned int coarse_cell_idx = p4est_tree_to_coarse_cell_permutation[idx];
+        lvl_it coarse_cell (&triangulation, 0, coarse_cell_idx);
+        set_subdomain_id_in_zorder_recursively(coarse_cell,
+                                               current_proc_idx,
+                                               current_cell_idx,
+                                               n_active_cellls,
+                                               n_partitions);
+      }
+
+    // ensure terminal siblings are on the same
```
Comment (tjhei) [inline source/grid/grid_tools.cc]: can you expand this to say:
If all children of a cell are active (so we have a cell that is refined once and no part is refined further), p4est places all of them on the same processor. The new owner will be the processor with the largest number of children (ties are broken by picking the lower rank). Duplicate this logic here.

File: source/grid/grid_tools.cc [hunk 13]
```diff
@@ -2125,6 +2191,155 @@ namespace GridTools
...
+  template <int dim, int spacedim>
+  void
+  partition_multigrid_levels (Triangulation<dim,spacedim> &triangulation)
+  {
+    unsigned int n_levels = triangulation.n_levels();
+    for (int lvl=n_levels-1; lvl>=0; --lvl)
+      {
+        typename Triangulation<dim,spacedim>::cell_iterator
+        cell = triangulation.begin(lvl),
+        endc=triangulation.end(lvl);
```
Comment (tjhei) [inline source/grid/grid_tools.cc]: spaces around ``=``

File: tests/sharedtria/tria_custom_refine.mpirun=3.output [hunk 14]
```diff
@@ -0,0 +1,92 @@
+
+DEAL:0:2d::(CellId,subdomain_id) for each active cell:
```
Comment (tjhei) [inline tests/sharedtria/tria_custom_refine.mpirun=3.output]: These are only the ``owned`` ones, though? Maybe print also the ghost cells?

File: tests/sharedtria/tria_multigrid_01.cc [hunk 15]
```diff
@@ -0,0 +1,172 @@
...
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE at
+// the top level of the deal.II distribution.
+//
+// ---------------------------------------------------------------------
+
+
+// create a shared tria mesh and distribute multigrid
```
Comment (tjhei) [inline tests/sharedtria/tria_multigrid_01.cc]: " in z-order and compare partition against p4est"

File: tests/sharedtria/tria_zorder_01.mpirun=3.output [hunk 16]
```diff
@@ -0,0 +1,8 @@
+
```
Comment (tjhei) [inline tests/sharedtria/tria_zorder_01.mpirun=3.output]: please rename this (and the other output files that need p4est) to contain ``.with_p4est=true.``

File: tests/sharedtria/tria_zorder_02.cc [hunk 17]
```diff
@@ -0,0 +1,83 @@
...
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE at
+// the top level of the deal.II distribution.
+//
+// ---------------------------------------------------------------------
+
+
+// create a shared tria mesh and distribute with zorder scheme
+// Unlike tria_zorder_01, this test does not need p4est library
```
Comment (tjhei) [inline tests/sharedtria/tria_zorder_02.cc]: Can you run the tests that don't need p4est also with dim=1 please? I am curious if that works. ;-)

Comment (tcclevenger) [inline tests/sharedtria/tria_zorder_02.cc]: We get the following error. I went ahead and changed back to only 2 and 3 dim. 

An error occurred in line <804> of file </home/tcclevenger/dealii/deal-git/dealii/include/deal.II/base/tensor.h> in function
    ArrayElementType& dealii::internal::TensorSubscriptor::subscript(ArrayElementType*, unsigned int, dealii::internal::int2type<dim>) [with ArrayElementType = dealii::Tensor<0, 1, double>; int dim = 1]
The violated condition was: 
    i<dim
Additional information: 
    Index 1 is not in the half-open range [0,1).

File: source/grid/grid_tools.cc (hunk 3 above)
Comment (tcclevenger) [inline source/grid/grid_tools.cc]: This was the only change I was a bit unsure on. I added the function <tt>compute_cell_halo_layer_on_level()</tt> as an internal function in <tt>shared_tria.cc</tt> directly above the function <tt>partition()</tt>. Let me know if I have misunderstood you and you had something else in mind.

File: source/distributed/shared_tria.cc (hunk 2 above)
Comment (davydden) [inline source/distributed/shared_tria.cc]: > I think you are missing an else before AssertThrow, Denis.

sure.

> we don't want to do the multigrid assignment for partition_custom_signal.

we need an Assert for that, probably already there...

**Comment by davydden** [comment]
@tcclevenger

> I added the function compute_cell_halo_layer_on_level() as an internal function in shared_tria.cc directly above the function partition(). Let me know if I have misunderstood you and you had something else in mind.

indeed, I was thinking about something slightly different.

The original placement of the function inside `GridTools` IMO is ok. Someone may be using it for something else at some point, so i would keep it there. That's why i suggested to use a general predicated, similar to https://www.dealii.org/developer/doxygen/deal.II/namespaceGridTools.html#a7ed62cdc6910a2ae39e8caa3142c1ad1


Second thought was that your function and `compute_active_cell_halo_layer` are almost identical. The only difference being in the loop:
```
for (typename MeshType::active_cell_iterator
       cell = mesh.begin_active();
       cell != mesh.end(); ++cell)
```
vs
```
for (typename parallel::shared::Triangulation<dim,spacedim>::cell_iterator
       cell = tria.begin(level);
       cell != tria.end(level); ++cell)
```
One could try extending the existing `compute_active_cell_halo_layer` to accept another argument, i.e. `const unsigned int level = numbers::invalid_unsigned_int` (the default means do it on the whole mesh), but given the fact that the two functions are so small, it probably does not really make sense here.

So bottom line is, please move it back to `GridTools` with `template <class MeshType>` and use predicate (as you already do).

File: tests/sharedtria/tria_zorder_02.cc (hunk 17 above)
Comment (bangerth) [inline tests/sharedtria/tria_zorder_02.cc]: That seems like the wrong approach. If you get an error, find why we hit it rather than not test the function that triggers it :-)

Comment (davydden) [inline tests/sharedtria/tria_zorder_02.cc]: running in debugger (gdb/lldb) and doing backtrace most likely immediately give you an idea what is wrong.

Comment (tcclevenger) [inline tests/sharedtria/tria_zorder_02.cc]: Ah, Sorry I was under the impression that it was expected to fail for something unrelated. 

Yes, it was only because during refinement I tried to access <tt>cell->center[1] </tt> when of course we were only in 1D so there was no 2nd coordinate value. Should I change up my test to cover for this or just leave it at 2D and 3D?

Comment (bangerth) [inline tests/sharedtria/tria_zorder_02.cc]: If it is easy to adjust the test, then it's worthwhile doing because that ensures that the 1d case is also tested. In general, the more test coverage you can get, the better.

File: source/distributed/shared_tria.cc (hunk 2 above)
Comment (tjhei) [inline source/distributed/shared_tria.cc]: > we need an Assert for that, probably already there...

I don't understand. We don't want to call ``partition_multigrid_levels`` if the partitioning method is set to custom. This is implemented correctly in the current version of the PR. What do you want to assert?

**Comment by davydden** [comment]
@tjhei 

> I don't understand. We don't want to call partition_multigrid_levels if the partitioning method is set to custom. This is implemented correctly in the current version of the PR. What do you want to assert?

I don't think it is implemented correctly in the current state. The current code has
```
if (settings & partition_metis)
 +        {
 +          dealii::GridTools::partition_triangulation (this->n_subdomains, *this);
 +        }
 +      else if (settings & partition_zorder)
 +        {
 +          dealii::GridTools::partition_triangulation_zorder (this->n_subdomains, *this);
 +        }
 +      else if (settings & partition_custom_signal)
 +        {
 +          // User partitions mesh manually
 +        }
 +      else
 +        {
 +          AssertThrow(false, ExcInternalError())
 +        }
 +
 +      // custom partition require manual partitioning of level cells
 +      if ((settings & construct_multigrid_hierarchy) && !(settings & partition_custom_signal))
 +        dealii::GridTools::partition_multigrid_levels(*this);
```
and thereby it is allowed to provide `settings` which contain incompatible `construct_multigrid_hierarchy` and `partition_custom_signal`. That's why i think a better way to do it is

```
if (settings & partition_metis)
 +        {
 +          dealii::GridTools::partition_triangulation (this->n_subdomains, *this);
 +        }
 +      else if (settings & partition_zorder)
 +        {
 +          dealii::GridTools::partition_triangulation_zorder (this->n_subdomains, *this);
 +        }
 +      else if (settings & partition_custom_signal)
 +        {
 +           Assert (!(settings &construct_multigrid_hierarchy), ExcMessage());
 +          // User partitions mesh manually
 +        }
 +      else
 +        {
 +          AssertThrow(false, ExcInternalError())
 +        }
 +
 +      // custom partition require manual partitioning of level cells
 +      if (settings & construct_multigrid_hierarchy)
 +        dealii::GridTools::partition_multigrid_levels(*this);
```

File: source/distributed/shared_tria.cc [hunk 18]
```diff
@@ -76,14 +146,84 @@ namespace parallel
                   active_halo_layer.find(cell) == active_halo_layer.end())
                 cell->set_subdomain_id(numbers::artificial_subdomain_id);
             }
+
+          // loop over all cells in multigrid hierarchy and mark artificial:
+          if (settings & construct_multigrid_hierarchy)
+            {
+              std_cxx11::function<bool (const typename parallel::shared::Triangulation<dim,spacedim>::cell_iterator &)> predicate
```
Comment (tcclevenger) [inline source/distributed/shared_tria.cc]: @davydden I think that the flag <tt>construct_multigrid_hierarchy</tt> is still useful for a computation where the mesh is distributed manually, we only don't want to call the function <tt>GridTools::partition_multigrid_levels</tt> as the user must manual distribute the level cells. 

For example here we would like to be able to use the flag  <tt>construct_multigrid_hierarchy</tt> alongside <tt>partition_custom_signal</tt> to mark artificial cells.

Comment (davydden) [inline source/distributed/shared_tria.cc]: i see, i did not notice this usage of the flag and had an impression that you only need it in the constructor.

File: include/deal.II/distributed/shared_tria.h [hunk 19]
```diff
@@ -150,6 +207,12 @@ namespace parallel
       bool with_artificial_cells() const;
 
     private:
+
+      /**
+       * Settings
+       */
+      Settings settings;
```
Comment (davydden) [inline include/deal.II/distributed/shared_tria.h]: I presume settings won't be altered? then it's `const Settings settings`.

File: source/distributed/shared_tria.cc [hunk 20]
```diff
@@ -76,14 +103,85 @@ namespace parallel
                   active_halo_layer.find(cell) == active_halo_layer.end())
                 cell->set_subdomain_id(numbers::artificial_subdomain_id);
             }
+
+          // loop over all cells in multigrid hierarchy and mark artificial:
+          if (settings & construct_multigrid_hierarchy)
```
Comment (davydden) [inline source/distributed/shared_tria.cc]: it appears that `construct_multigrid_hierarchy` only works when `allow_artificial_cells==true`? If so, we need an Assert in the constructor.

Comment (tjhei) [inline source/distributed/shared_tria.cc]: It should be possible to make it work without that setting, but I don't see a good reason for allowing that. ``allow_artificial`` is more like a backwards-compatibility thing, right?

File: source/distributed/shared_tria.cc (hunk 18 above)
Comment (tjhei) [inline source/distributed/shared_tria.cc]: Oh, now I get it. Your point is that the flag ``construct_multigrid_hierarchy`` is redundant when using custom partitioning. While this is true, I have two arguments to keep things as is:
1. If I want to switch from metis to custom in a particular program, I would also need to fiddle with the settings.
2. I can envision a reusable "custom" routine that looks at the settings to decide if multigrid cells need to be partitioned or not.

File: source/distributed/shared_tria.cc (hunk 20 above)
Comment (davydden) [inline source/distributed/shared_tria.cc]: > backwards-compatibility thing, right?

not really, it's a feature, see https://github.com/dealii/dealii/issues/1368#issuecomment-131360821

the reason for that auxiliary vector to store ids is due to the problem here https://github.com/dealii/dealii/pull/1674#issuecomment-143314553

Comment (davydden) [inline source/distributed/shared_tria.cc]: but since this was a few years ago, i don't remember exactly all the details.

Comment (tcclevenger) [inline source/distributed/shared_tria.cc]: Did we reach a verdict on whether or not to add the assert?

Comment (davydden) [inline source/distributed/shared_tria.cc]: i would add assert because in the current state of code `construct_multigrid_hierarchy` won't be used in full unless you have `allow_artificial_cells==true`.

**Comment by bangerth** [comment]
Let's postpone this until after 8.5.

**Comment by tjhei** [comment]
> Let's postpone this until after 8.5.

good to merge?

**Comment by tjhei** [comment]
@tcclevenger your branch doesn't compile with 64bit indices, can you take a look?
See https://cdash.kyomu.43-1.org/viewBuildError.php?buildid=6253
You need to configure your deal.II with ``-D DEAL_II_WITH_64BIT_INDICES=ON`` and recompile.

**Comment by tjhei** [comment]
@tcclevenger it looks like @kronbichler beat you: #4142

**Comment by tcclevenger** [comment]
@tjhei This would explain why it is compiling with no errors this morning. haha I was very confused. Thanks @kronbichler !

Also, I found a bug. I was not making sure that a proc knew about its parent's cells when setting level subdomain ids to artificial.  My tests did not catch this since the meshes were subdivided hypercubes so there was more than one cell on the coarsest level and all processors were present. Does this need to be fixed in a separate pull request or can I commit to this branch even though it is merged?

**Comment by tjhei** [comment]
> Does this need to be fixed in a separate pull request

Yes.

**Comment by bangerth** [comment]
Yes, separate pull request that will be separately reviewed then.

Since you already found a bug, can you make sure that a testcase for this bug becomes part of the separate pull request? Little programs that demonstrate a bug make for excellent testcases.

File: source/distributed/shared_tria.cc [hunk 21]
```diff
@@ -76,14 +107,85 @@ namespace parallel
...
+        for (; cell!=endc; ++cell)
+          if (cell->is_locally_owned())
+            n_my_cells += 1;
+
+        unsigned int total_cells;
+        int ierr = MPI_Allreduce (&n_my_cells, &total_cells, 1,
+                                  MPI_UNSIGNED, MPI_SUM, this->get_communicator());
+        AssertThrowMPI(ierr);
+
+        Assert(total_cells == this->n_active_cells(),
```
Comment (davydden) [inline source/distributed/shared_tria.cc]: i hit this Assert, i think it's wrong as `total_cells` is sum of all locally owned, whereas `n_active_cells()` is a MPI-local value. 

But there is another issue: `n_global_active_cells` and `n_active_cells` uses `number_cache`, but this case is update in `execute_coarsening_and_refinement()` **after** `partition()`. I see if I can fix this...

Comment (bangerth) [inline source/distributed/shared_tria.cc]: The code also seems unnecessary -- there is already `Triangulation::n_globally_active_cells`.

Comment (tjhei) [inline source/distributed/shared_tria.cc]: > i hit this Assert, i think it's wrong as total_cells is sum of all locally owned, whereas n_active_cells() is a MPI-local value.

I am confused, ``total_cells`` is the sum of ``is_locally_owned`` over all ranks, which should be equal to ``n_active_cells`` (number of cells that have no children).

> ``n_active_cells`` uses ``number_cache``

Really? I thought only the global one does.

Comment (tjhei) [inline source/distributed/shared_tria.cc]: Can you create a (failing) test?

Comment (tjhei) [inline source/distributed/shared_tria.cc]: > The code also seems unnecessary -- there is already Triangulation::n_globally_active_cells.

I disagree. This checks that the user created a consistent partition. This is especially useful for a user-defined partitioning method.

Comment (davydden) [inline source/distributed/shared_tria.cc]: > I am confused, total_cells is the sum of is_locally_owned over all ranks, which should be equal to n_active_cells

no, it is not! `n_active_cells` is the number of active cells **on this MPI core**, not over all cores.

Comment (davydden) [inline source/distributed/shared_tria.cc]: > Really? I thought only the global one does.

yes, have a look yourself https://www.dealii.org/developer/doxygen/deal.II/grid_2tria_8cc_source.html#l11244

Comment (davydden) [inline source/distributed/shared_tria.cc]: > Can you create a (failing) test?

i will, but later...

Comment (bangerth) [inline source/distributed/shared_tria.cc]: Oh, yes, I get it -- every processor of course has *all* cells, but here every processor counts how many cells it owns itself (at the end of the partitioning step) and that should of course equal the total number of cells (which for the case of a `p::shared::Tria` is equal to both `n_active_cells` and `n_globally_active_cells`).

So I'm ok with it. No change needed for me here.

Comment (tjhei) [inline source/distributed/shared_tria.cc]: > no, it is not! n_active_cells is the number of active cells on this MPI core, not over all cores.

Which is equal to the total number of cells with no children (and includes owned, ghost, and artificial). For a shared triangulation we have ``n_active_cells=n_globally_active_cells``.

Could it be that the only problem is that the number cache is wrong?

Comment (davydden) [inline source/distributed/shared_tria.cc]: see https://github.com/dealii/dealii/pull/4167
